jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/ServletContextHandler.java

```diff
@@ -2829,10 +2829,6 @@ public InputStream getResourceAsStream(String path)
         @Override
         public Set<String> getResourcePaths(String path)
         {
-            // This is an API call from the application which may pass non-canonical paths.
-            // Thus, we canonicalize here, to avoid the enforcement of canonical paths in
-            // ContextHandler.this.getResource(path).
-            path = URIUtil.canonicalPath(path);
```

Comment:
This was a very deliberate design to canonicalize paths provided via the servlet API.    I'd be very cautious about any changes here and definitely re-run the TCK tests.

Reply:
In Jetty 10/11 we have the exact same comment and code.... but I think we have changed the meaning of the `canonicalPath` method. In 11 the javadoc for that method says:
```java
    /**
     * Convert a decoded URI path to a canonical form.
     * <p>
     * All segments of "." and ".." are factored out.
     * Null is returned if the path tries to .. above its root.
     * </p>
     *
     * @param path the decoded URI path to convert. Any special characters (e.g. '?', "#") are assumed to be part of
     * the path segments.
     * @return the canonical path, or null if path traversal above root.
     * @see #canonicalURI(String)
     */
```

whilst in 12 is says:
```java
    /**
     * Canonicalize a URI path to a form that is unambiguous and safe to use with the JVM {@link URI} class.
     * <p>
     * Decode only the safe characters in a URI path and strip parameters of UTF-8 path.
     * Safe characters are ones that are not special delimiters and that can be passed to the JVM {@link URI} class.
     * Unsafe characters, other than '{@code /}' will be encoded.  Encodings will be uppercase hex.
     * Canonical paths are also normalized and may be used in string comparisons with other canonical paths.
     * <p>
     * For example the path {@code /fo %2fo/b%61r} will be normalized to {@code /fo%20%2Fo/bar},
     * whilst {@link #decodePath(String)} would result in the ambiguous and URI illegal {@code /fo /o/bar}.
     * @param encodedPath An encoded URI path
     * @return the canonical path or null if it is non-normal
     * @see #decodePath(String)
     * @see #normalizePath(String)
     * @see URI
     */
```

So it used to take a decoded path, but now it takes an encoded path.

Thus perhaps we need a new method to canonicalize (or at least normalize dot segments etc.) decoded paths.

This is the old problem that the servlet spec really never specified what space these strings are in.   I think they expect "/./" and "/../" segments to be handled, but not for % encodings to be expanded.  But the stupid thing is that dot segments are meant to be resolved before % encoding according to the RFCs, so it makes no sense to normalize dot segments and then ignore % encodings... but that's what we used to do... so I guess we should keep doing it for servlet API methods